libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using prompts = Microsoft.Bot.Builder.Prompts;
+using static Microsoft.Bot.Builder.Prompts.PromptValidatorEx;
+using Microsoft.Bot.Schema;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Prompts;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    public class OAuthPromptSettingsWithTimeout : prompts.OAuthPromptSettings
+    {
+        public int? Timeout { get; set; }
+    }
+
+    public class OAuthPromptState : PromptOptions
+    {
+        public OAuthPromptState(PromptOptions defaultPromptOptions) : base()
+        {
+            if (defaultPromptOptions != null)
+            {
+                PromptString = PromptString ?? defaultPromptOptions.PromptString;
+                PromptActivity = PromptActivity ?? defaultPromptOptions.PromptActivity;
+                Speak = Speak ?? defaultPromptOptions.Speak;
+                RetryPromptString = RetryPromptString ?? defaultPromptOptions.RetryPromptString;
+                RetryPromptActivity = RetryPromptActivity ?? defaultPromptOptions.RetryPromptActivity;
+                RetrySpeak = RetrySpeak ?? defaultPromptOptions.RetrySpeak;
+            }
+        }
+
+        public DateTime Expires { get; set; }
+    }
+
+    public class OAuthPrompt : Control, IDialog, IDialogContinue
+    {
+        private prompts.OAuthPrompt _prompt;
+        private OAuthPromptSettingsWithTimeout _settings;
+
+        public OAuthPrompt(OAuthPromptSettingsWithTimeout settings, PromptValidator<TokenResult> validator = null)
+        {
+            _settings = settings ?? throw new ArgumentException(nameof(settings));
+            _prompt = new prompts.OAuthPrompt(settings, validator);
+        }
+
+        public async Task DialogBegin(DialogContext dc, object dialogArgs = null)
+        {
+            if (dc == null)
+                throw new ArgumentNullException(nameof(dc));
+            if (dialogArgs == null)
+                throw new ArgumentNullException(nameof(dialogArgs));
+
+            var promptOptions = (PromptOptions)dialogArgs;
```

Comment:
This is odd. An object was passed in. Could be anything. 

Then, for reasons that are unclear, you're *hoping* it's a PromptOptions class and casting it. The resulting exception will be unexpeceted and not very helpful .

The correct answer seems to be:
1. Pass in a PromptOptions rather than an object. 
2. If #1 is impossible for some reason, at least check the type: if (! dialogArge is PromptOptions) throw new ArgumentException...

Reply:
the argument passed in is an object because the funciton is an implementation of the idialog interface, so i can't change it. i have added a cast check and am throwing an argument exception now as suggested in 2 above.

---
In reply to: [184818971](https://github.com/Microsoft/botbuilder-dotnet/pull/520#discussion_r184818971) [](ancestors = 184818971)